### DISCUSSION TITLE
SaveImage with Compose

### DISCUSSION THREAD

**Discussion by vikashg** [Project-MONAI/MONAI#6036] [Q&A]
Hi all, 
I want to do some preprocessing and save the preprocessed images but if I write a transformation chain as 
```
transforms = Compose([LoadImage(), EnsureChannelFirst(), Resize(spatial_size = (nx, ny, nz)), 
SaveImage(output_dir = './ResizedImage', output_ext='.nii.gz')])
```
If I save this image I see that the image saved has its original dimensions. I wonder if there is something that I am doing wrong.

**Comment by KumoLiu** [comment]
Hi @vikashg, yes `SaveImage` will resample the image by default, you could set `resample=False` then it will not resample the image based on the `spatial_shape` (and `original_affine`). 
https://github.com/Project-MONAI/MONAI/blob/66422f805326a1034e040a9eacc2c8220717052e/monai/transforms/io/array.py#L320-L321

Hope it can help you, thanks!

**Reply by vikashg** [reply]
I tried using both `resample=True` and `resample=False` and they both yield the same result, i.e., to save the image in the original dimensions. However, if I break the transform list to 
`transforms = Compose([LoadImage(), EnsureChannelFirst(), Resize(spatial_size = (nx, ny, nz))])`
the output has the resampled shape which can be saved using a `ITKWriter`. Here is the working code 
```
transform_chain = Compose([LoadImaged(keys=["image"]), EnsureChannelFirstd(keys=["image"]), 
Resized(keys=["image"], spatial_size=(-1, -1, 108))])
writer = ITKWriter()

a = transform_chain(image_filename)
a['image_meta_dict']['spatial_shape'] = [512, 512, 108]

writer.set_data_array(a['image'])
writer.set_metadata(a['image_meta_dict'])
writer.write(out_filename)
```
What doesn't work is 

```
transform_chain = Compose([LoadImaged(keys=["image"]), EnsureChannelFirstd(keys=["image"]), 
Resized(keys=["image"], spatial_size=(-1, -1, 108)) , SaveImaged(keys=['image], output_dir='./out_dir', output_postfix='resized', 
output_ext = '.nii.gz', resample=True)])
```
I am using monai 1.1.0

**Reply by KumoLiu** [reply]
Hi @vikashg, could you please share the data you were using if possible? I will try the code and take a deep look at it. Thanks!

**Reply by vikashg** [reply]
Here is the [image](https://drive.google.com/file/d/1-mXgCfzMxEiGMGdQxagjwoam_r3AnKUY/view?usp=sharing)

**Reply by vikashg** [reply]
And thanks for taking a look

**Reply by KumoLiu** [reply]
Hi @vikashg, I tried your data and if set `resample=False`, it works.
```
from monai.transforms import LoadImaged, EnsureChannelFirstd, Compose, Resized, SaveImaged

data_path = "/workspace/Code/MONAI/output/002D_raw.nii.gz"

data = {"image": data_path}

transform_chain = Compose([
    LoadImaged(keys=["image"]), 
    EnsureChannelFirstd(keys=["image"]), 
    Resized(keys=["image"], spatial_size=(-1, -1, 108)),
    SaveImaged(keys=["image"], output_dir='.', output_postfix='resized', output_ext = '.nii.gz', resample=False)
            ])
out_resized = transform_chain(data)


out = LoadImaged(keys=["image"])({"image": "./002D_raw/002D_raw_resized.nii.gz"})
print(out['image'].shape)  # --> torch.Size([512, 512, 108])
```

**Reply by vikashg** [reply]
ok Let me try 
Thanks 
Vikash

**Reply by vikashg** [reply]
Ya it works now, I dont know what I was doing wrong :( Sorry about this. It works 
Thanks for the help.

**Reply by wyli** [reply]
it seems `resample=False` is a better default for `SaveImage`... I'm creating a feature request to update the default

**Reply by vikashg** [reply]
Yes, thanks @wyli

**Reply by vikashg** [reply]
@KumoLiu I think I spoke too soon. Can you try one more time with a slice number other than 108 and see if it works? Because the image I shared with you already has 108 slices. So, it was a bad test case. 

Thanks 
Vikash

**Reply by KumoLiu** [reply]
Hi @vikashg, I just take a look at the original data, the shape is (512, 512, 83). I would like to say you may check the data after transform chain, it didn't replace the data that was originally with key “image”.
```
from monai.transforms import LoadImage

data = LoadImage(image_only=True)(data_path)
print(data.shape)
```
Thanks!

**Reply by vikashg** [reply]
So @KumoLiu. Here is the code (I just added 1 line before the transformation to get the shape of the input)
```
from monai.transforms import LoadImaged, EnsureChannelFirstd, Compose, Resized, SaveImaged

data_path = "/workspace/app/002D_raw.nii.gz"

data = {"image": data_path}

input1 = LoadImaged(keys=['image'])(data)
print(input1['image'].shape)

transform_chain = Compose([
    LoadImaged(keys=["image"]),
    EnsureChannelFirstd(keys=["image"]),
    Resized(keys=["image"], spatial_size=(-1, -1, 48)),
    SaveImaged(keys=["image"], output_dir='.', output_postfix='resized', output_ext = '.nii.gz', resample=True)
            ])
out_resized = transform_chain(data)


out = LoadImaged(keys=["image"])({"image": "./002D_raw/002D_raw_resized.nii.gz"})
print(out['image'].shape)  # --> torch.Size([512, 512, 108])
```
and here is the output
```
torch.Size([512, 512, 83])
2023-02-23 02:55:58,101 INFO image_writer.py:194 - writing: 002D_raw/002D_raw_resized.nii.gz
torch.Size([512, 512, 83])
```
So the data has the same size after resampling

**Reply by KumoLiu** [reply]
Hi @vikashg, l noticed that you still using `resample=True`. If you don't want data resample to the original shape, you may need set it to `False`.

**Reply by vikashg** [reply]
Aah ok. I will give it a go tomorrow morning thanks
